### PR TITLE
fix(core): `toJsonString()` cannot handle list intrinsics

### LINKED ISSUE #13465: (core): Templating error on custom-resource introduced with version 1.92
````typescript
for (let index = 0; index < existingVpc.availabilityZones.length; index++) {
      let getEndpointIp = new AwsCustomResource(this, idPrefix + 'GetEndpointIp' + index, {
        policy: AwsCustomResourcePolicy.fromSdkCalls({resources: AwsCustomResourcePolicy.ANY_RESOURCE}),
        onUpdate: {
          action: 'describeNetworkInterfaces',
          service: 'EC2',
          parameters: {NetworkInterfaceIds: vpcendpoint.vpcEndpointNetworkInterfaceIds},
          physicalResourceId: PhysicalResourceId.of('DirectoryId'),
          outputPath: 'NetworkInterfaces.' + index + '.PrivateIpAddress'
        }
      })
      targetGroup.addTarget(new IpTarget(Token.asString(getEndpointIp.getResponseField('NetworkInterfaces.' + index + '.PrivateIpAddress'))));
    }
````

We are doing this, because we need to put an api gateway behind a load balancer (company policy). If there is a better solution to do this, feedback is welcome.

### Reproduction Steps

this is the output with aws-cdk/core 1.91

````yaml
...
  announceGetEndpointIp0D5B4E67A:
    Type: Custom::AWS
    Properties:
      ServiceToken:
        Fn::GetAtt:
          - AWS12345678
          - Arn
      Create:
        action: describeNetworkInterfaces
        service: EC2
        parameters:
          NetworkInterfaceIds:
            Fn::GetAtt:
              - announcevpcannouncevpcendpoint55991BB5
              - NetworkInterfaceIds
        physicalResourceId:
          id: DirectoryId
        outputPath: NetworkInterfaces.0.PrivateIpAddress
      Update:
        action: describeNetworkInterfaces
        service: EC2
        parameters:
          NetworkInterfaceIds:
            Fn::GetAtt:
              - announcevpcannouncevpcendpoint55991BB5
              - NetworkInterfaceIds
        physicalResourceId:
          id: DirectoryId
        outputPath: NetworkInterfaces.0.PrivateIpAddress
      InstallLatestAwsSdk: true
    DependsOn:
      - announceGetEndpointIp0CustomResourcePolicyC5EFFE47
    UpdateReplacePolicy: Delete
    DeletionPolicy: Delete
    Metadata:
      aws:cdk:path: announce-dev-local/announce-GetEndpointIp0/Resource/Default
...
````

this is the output with aws-cdk/core 1.92

```yaml
...
  announceGetEndpointIp0D5B4E67A:
    Type: Custom::AWS
    Properties:
      ServiceToken:
        Fn::GetAtt:
          - AWS12345678
          - Arn
      Create:
        Fn::Join:
          - ""
          - - '{"action":"describeNetworkInterfaces","service":"EC2","parameters":{"NetworkInterfaceIds":"'
            - Fn::GetAtt:
                - announcevpcannouncevpcendpoint55991BB5
                - NetworkInterfaceIds
            - '"},"physicalResourceId":{"id":"DirectoryId"},"outputPath":"NetworkInterfaces.0.PrivateIpAddress"}'
      Update:
        Fn::Join:
          - ""
          - - '{"action":"describeNetworkInterfaces","service":"EC2","parameters":{"NetworkInterfaceIds":"'
            - Fn::GetAtt:
                - announcevpcannouncevpcendpoint55991BB5
                - NetworkInterfaceIds
            - '"},"physicalResourceId":{"id":"DirectoryId"},"outputPath":"NetworkInterfaces.0.PrivateIpAddress"}'
      InstallLatestAwsSdk: true
    DependsOn:
      - announceGetEndpointIp0CustomResourcePolicyC5EFFE47
    UpdateReplacePolicy: Delete
    DeletionPolicy: Delete
    Metadata:
      aws:cdk:path: announce-dev-local/announce-GetEndpointIp0/Resource/Default
...
```

### What did you expect to happen?

After using `cdk deploy` we expected the same behavior as in version 1.91 with successful deployment.

### What actually happened?

The error only occurs during a `cdk deploy` with version 1.92.

```sh
26/44 | 8:32:42 AM | CREATE_FAILED        | Custom::AWS                               | announce-GetEndpointIp2/Resource/Default (announceGetEndpointIp20B447BE7) Template error: every Fn::Join object requires two parameters, (1) a string delimiter and (2) a list of strings to be joined or a function that returns a list of strings (such as Fn::GetAZs) to be joined.
	new CustomResource (/builds/marceldarvis/toolbox/services/announce-pipeline-modules-service/node_modules/@aws-cdk/core/lib/custom-resource.ts:119:21)
	\_ new AwsCustomResource (/builds/marceldarvis/toolbox/services/announce-pipeline-modules-service/node_modules/@aws-cdk/custom-resources/lib/aws-custom-resource/aws-custom-resource.ts:374:27)
```

### Environment

  - **CDK CLI Version  :1.92**
  - **Framework Version:None**
  - **Node.js Version:v14.16.0** 
  - **OS               :linux**
  - **Language (Version):TypeScript (4.2.3)** 

### Other

We assume that this issue is related to this change:

- https://github.com/aws/aws-cdk/commit/3cb31043a42b035f6dcd2a318836d4bfc4973151

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by rix0rrr**
The previous attempt at a fix missed one important case: the types
of the values involved in the `{ Fn::Join }` expression didn't actually
match up. They all needed to be strings, but the previous implentation
just dropped list-typed values in there.

Unfortunately, there is no way to do it correctly with just string
manipulation in CloudFormation (`{ Fn::Join }` etc), so we'll have
to resort to using a Custom Resource if we encounter list values.

Actually fixes #13465.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/core/lib/private/cloudformation-lang.ts [hunk 1]
```diff
@@ -211,12 +214,34 @@ function tokenAwareStringify(root: any, space: number, ctx: IResolveContext) {
...
+        //
+        // However, STRINGIFY would need to execute at CloudFormation deployment time, and that doesn't exist.
+        //
+        // We could *ALMOST* use:
+        //
+        //   '{"listValue":["' ++ JOIN('","', { Ref: MyList }) ++ '"]}'
+        //
+        // But that has the unfortunate side effect that if `CFN_EVAL({ Ref: MyList }) == []`, then it would
+        // evaluate to `[""]`, which is a different value. Since CloudFormation does not have arbitrary
+        // conditionals there's no way to deal with this case properly.
```
Comment (jogold) [inline packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:232]: @rix0rrr  I think I found a way with a `Fn::Join` on a `Fn::Split` acting like a replace:


```ts
cdk.Fn.join('', cdk.Fn.split('""', myString)) // acts like a replace
```


For #13465 and the issue with the array of ENIs, it looks like that

```ts
[cdk.Token.asList(cdk.Fn.join('', cdk.Fn.split('""', cdk.Stack.of(stack).toJsonString(cdk.Fn.join('","', endpoint.vpcEndpointNetworkInterfaceIds)))))]
```

Gives the following CF
```json
{
  "Fn::Join": [
    "",
    [
      "{\"action\":\"describeNetworkInterfaces\",\"service\":\"EC2\",\"parameters\":{\"NetworkInterfaceIds\":[",
      {
        "Fn::Join": [
          "",
          {
            "Fn::Split": [
              "\"\"",
              {
                "Fn::Join": [
                  "",
                  [
                    "\"",
                    {
                      "Fn::Join": [
                        "\\\",\\\"",
                        {
                          "Fn::GetAtt": [
                            "OtherEndpoint88C37F72",
                            "NetworkInterfaceIds"
                          ]
                        }
                      ]
                    },
                    "\""
                  ]
                ]
              }
            ]
          }
        ]
      },
      "]},\"physicalResourceId\":{\"id\":\"physical-id\"},\"outputPath\":\"NetworkInterfaces.0.PrivateIpAddress\"}"
    ]
  ]
}
```

Successfully tested this with a `GatewayVpcEndpoint` which has no ENIs and hence returns an empty list for the `vpcEndpointNetworkInterfaceIds` attribute.

WDYT?

Comment (rix0rrr) [reply packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:232]: EHRMAGERD

Comment (rix0rrr) [reply packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:232]: You are like a wizard.

A dirty wizard...

Comment (rix0rrr) [reply packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:232]: I like it though.

Comment (rix0rrr) [reply packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:232]: Wait no. This solution will break when presented with `[""]`.

That value would successfully make it through the transformation and then be left as `[]`.

Comment (jogold) [reply packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:232]: ahh 😄 

You're right, there's may be something to do if we first join the array with something else than `","`... I need to think about it 🤔

Comment (jogold) [reply packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:232]: Bu then maybe the CR is the right solution

Comment (rix0rrr) [reply packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:232]: I'm also thinking about it... afraid nothing's to be done, as:

```
JOIN(<whatever>, []) == JOIN(<whatever>, [""]) == ""
```

Once we've JOINed, we've lost the ability to make the distinction between `[]` and `[""]`... right? 

(I'm starting to doubt myself now)

Comment (jogold) [reply packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:232]: I don't think there's a way to do it. You loose information after the first join:

```ts
[].join('","') === [''].join('","')
``` 

and the only "first operation" you can do on the array with CF is a `Fn::Join`.

Comment (jogold) [reply packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:232]: OK our comments crossed each other... I don't think it's possible in pure CF. This is sad 😢 .

Comment (rix0rrr) [reply packages/@aws-cdk/core/lib/private/cloudformation-lang.ts:232]: IF ONLY we had a functional `{ Fn::If }`...

**Review by njlynch — APPROVED** [review]

**Pull request merged**